File: week3/main.cpp
#include <iostream>

double f(double x)
{
	double xx = x * x;
	double xxx = xx * x;
	double xxxx = xxx * x;
	return 1 + 4 * x + 6 * xx + 4 * xxx + xxxx;
}

double TernarySearch(double (&funcRef)(double),
										 double left,
										 double right,
										 double eps=1e-10,
										 int maxIterations=1e6,
										 bool log=false)
{
	int iterations = 0;
	while (right - left > eps && iterations < maxIterations)
	{
		double ml = left * 2 / 3 + right / 3;
		double mr = left / 3 + right * 2 / 3;

		if (funcRef(ml) < funcRef(mr))
		{
			right = mr;
		}
		else
		{
			left = ml;
		}
		++iterations;
	}

	if (log)
	{
		std::cout << "TernarySearch iterations: " << iterations << std::endl << std::flush;
	}

	return (left + right) / 2;
}

int main(int argc, const char* argv[]) {
	double min = TernarySearch(f, -2.0, 2.0, 1e-15, 1e3, true);

	std::cout << min << ", " << f(min) << std::endl;

	return 0;
}

File: week3/main.rs
fn f(x: f64) -> f64 {
	let xx = x * x;
	let xxx = xx * x;
	let xxxx = xxx * x;
	1.0 + 4.0 * x + 6.0 * xx + 4.0 * xxx + xxxx
}

fn ternary_search(
	func: fn(f64) -> f64,
	left: f64,
	right: f64,
	eps: f64,
	max_iterations: i32) -> (f64, f64, i32) {

	let mut iterations = 0;
	let mut left = left;
	let mut right = right;

	while right - left > eps && iterations < max_iterations {
		let ml = left * 2.0 / 3.0 + right / 3.0;
		let mr = left / 3.0 + right * 2.0 / 3.0;

		if func(ml) < func(mr) {
			right = mr;
		} else {
			left = ml;
		}
		iterations += 1;
	}

	let r = left / 2.0 + right / 2.0;
	(r, func(r), iterations)
}

fn main() {
	let (min_at, min_val, iterations) = ternary_search(f, -2.0, 2.0, 1e-15, 1000);
	// let (min_at, min_val, iterations) = ternary_search(|x: f64| -> f64 {x * x}, -2.0, 2.0, 1e-15, 1000);
	println!("ternary_search iterations: {}", iterations);
	println!("min at: {}, value: {}", min_at, min_val);
}
